beacon-chain/core/blocks/deposit.go

```diff
@@ -146,7 +146,9 @@ func ProcessDeposit(beaconState *stateTrie.BeaconState, deposit *ethpb.Deposit,
 	pubKey := deposit.Data.PublicKey
 	amount := deposit.Data.Amount
 	index, ok := beaconState.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubKey))
-	if !ok {
+	// Also ensures that beacon state may not be the latest state hence `IndexByPubkey` may not always reflect to input state.
+	// Guard check using validator length.
+	if !ok || index >= uint64(len(beaconState.ValidatorsReadOnly())) {
```

Comment:
Isn't there a beaconState.ValidatorCount?

Reply:
```suggestion
	if !ok || index >= uint64(beaconState.NumValidators()) {
```

Reply:
Ah yes. Thanks for feedback!